src/adapter/src/copy_to.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Implementations around supporting the COPY TO protocol with the dataflow layer
+
+use std::collections::BTreeSet;
+
+use mz_adapter_types::connection::ConnectionId;
+use mz_compute_client::protocol::response::CopyToResponse;
+use mz_controller_types::ClusterId;
+use mz_repr::GlobalId;
+use mz_sql::session::user::User;
+use tokio::sync::oneshot;
+
+use crate::CopyToResponseUnary;
+
+/// A description of an active copy to from coord's perspective.
+#[derive(Debug)]
+pub struct ActiveCopyTo {
+    /// The user of the session that created the copy to.
+    pub user: User,
+    /// The connection id of the session that created the copy to.
+    pub conn_id: ConnectionId,
+    /// Sender to send responses to the client.
+    pub sender: Option<oneshot::Sender<CopyToResponseUnary>>,
+    /// The cluster that the copy to is running on.
+    pub cluster_id: ClusterId,
+    /// All `GlobalId`s that the copy to's expression depends on.
+    pub depends_on: BTreeSet<GlobalId>,
+    /// Whether we are already in the process of dropping the resources related to this copy to.
+    pub dropping: bool,
+}
+
+impl ActiveCopyTo {
+    /// Processes a copy to response.
+    pub(crate) fn process_response(&mut self, response: CopyToResponse) {
+        let resp = match response {
+            CopyToResponse::Rows(count) => CopyToResponseUnary::Rows(count),
+            CopyToResponse::Error(text) => CopyToResponseUnary::Error(text),
+            CopyToResponse::Canceled => CopyToResponseUnary::Canceled,
+        };
+        // Using an option around the sender so to get an owned value after take
+        // to call `send` on it.
+        if let Some(channel) = self.sender.take() {
+            let _ = channel.send(resp);
+        }
+    }
```

Comment:
See the comment in protocol.rs before reading this one!

Basically, I think you want to take the code in protocol.rs that aggregates up the `CopyToResponse`s and do that _here_. Instead of storing a special sender for `CopyToResponse`s in `ActiveCopyTo`, I think you just want to hang on to the `execute_ctx`, and then you can call `execute_ctx.retire(ExecuteResponse::Copied(row_count))` when you've seen all the `CopyToResponse`s from the dataflow layer that you expect. Does that make sense?

Reply:
@benesch This portion is where adapter is receiving a single and already aggregated CopyToResponse from the controller, and then passing it along to the oneshot receiver in `ExecuteResponse::Copying { rx ...` which was earlier returned after shipping the dataflow https://github.com/MaterializeInc/materialize/pull/25042/files#diff-9a169fe3ef85a53dbf876b6b69d8d4f6f6c78901afcdf2683f72066e924b9df2R1032-R1035 (and environmentd is await-ing on it below before returning to the user).

If I eventually want to send something like `ExecuteResponse::Copied(rows)` back, what should I return after shipping the dataflow then instead of `ExecuteResponse::Copying { rx...`? Because I think we still want to send something back to the user which blocks until we are done and we receive the final row count.

Reply:
Oh do you mean that we not return `ExecuteResponse::Copying { rx...` from sequencer but rather wait for the dataflow responses back there and directly return `ExecuteResponse::Copied`?

Reply:
Have updated the PR to send a `ExecuteResponse::Copied` directly.